pkg/controller/util/k8s/protocol.go

```diff
@@ -0,0 +1,236 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"encoding/json"
+	"fmt"
+
+	api "github.com/atomix/api/proto/atomix/controller"
+	"github.com/atomix/kubernetes-controller/pkg/apis/cloud/v1beta2"
+	storage "github.com/atomix/raft-storage-controller/pkg/apis/v1beta1"
+	"github.com/gogo/protobuf/jsonpb"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	apiPort      = 5678
+	protocolPort = 5679
+	probePort    = 5679
+)
+
+// NewClusterConfigMap returns a new ConfigMap for initializing Atomix clusters
+func NewClusterConfigMap(cluster *v1beta2.Cluster, storage *storage.RaftStorageClass, config interface{}) (*corev1.ConfigMap, error) {
+	clusterConfig, err := newNodeConfigString(cluster, storage)
+	if err != nil {
+		return nil, err
+	}
+
+	protocolConfig, err := newProtocolConfigString(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+			Labels:    cluster.Labels,
+		},
+		Data: map[string]string{
+			clusterConfigFile:  clusterConfig,
+			protocolConfigFile: protocolConfig,
+		},
+	}, nil
+}
+
+// getPodName returns the name of the pod for the given pod ID
+func getPodName(cluster *v1beta2.Cluster, pod int) string {
+	return fmt.Sprintf("%s-%d", cluster.Name, pod)
+}
+
+// getPodDNSName returns the fully qualified DNS name for the given pod ID
+func getPodDNSName(cluster *v1beta2.Cluster, pod int) string {
+	return fmt.Sprintf("%s-%d.%s.%s.svc.cluster.local", cluster.Name, pod, cluster.Name, cluster.Namespace)
+}
+
+// newNodeConfigString creates a node configuration string for the given cluster
+func newNodeConfigString(cluster *v1beta2.Cluster, storage *storage.RaftStorageClass) (string, error) {
+	members := make([]*api.MemberConfig, storage.Spec.Replicas)
+	for i := 0; i < int(storage.Spec.Replicas); i++ {
+		members[i] = &api.MemberConfig{
+			ID:           getPodName(cluster, i),
+			Host:         getPodDNSName(cluster, i),
+			ProtocolPort: protocolPort,
+			APIPort:      apiPort,
+		}
+	}
+
+	config := &api.ClusterConfig{
+		Members: members,
+	}
+
+	marshaller := jsonpb.Marshaler{}
+	return marshaller.MarshalToString(config)
+}
+
+// newProtocolConfigString creates a protocol configuration string for the given cluster and protocol
+func newProtocolConfigString(config interface{}) (string, error) {
+	bytes, err := json.Marshal(config)
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+// NewClusterDisruptionBudget returns a new pod disruption budget for the cluster group cluster
+func NewClusterDisruptionBudget(cluster *v1beta2.Cluster, storage *storage.RaftStorageClass) *policyv1beta1.PodDisruptionBudget {
```

Comment:
I haven't used this yet. Do we need to use it for this use case?

Reply:
We do need it, but we can skip it for the first iteration. I'd prefer to get some working version of this controller merged rather than seeking perfection on the first try.